base_layer/core/src/validation/block_validators/async_validator.rs

```diff
@@ -0,0 +1,432 @@
+            };
+
+            let mut coinbase_index = None;
+            let mut max_kernel_timelock = 0;
+            for (i, kernel) in kernels.iter().enumerate() {
+                kernel.verify_signature()?;
+
+                if kernel.is_coinbase() {
+                    if coinbase_index.is_some() {
+                        warn!(
+                            target: LOG_TARGET,
+                            "Block #{} failed to validate: more than one kernel coinbase", height
+                        );
+                        return Err(ValidationError::TransactionError(TransactionError::MoreThanOneCoinbase));
+                    }
+                    coinbase_index = Some(i);
+                }
+
+                max_kernel_timelock = cmp::max(max_kernel_timelock, kernel.lock_height);
+                kernel_sum.fees += kernel.fee;
+                kernel_sum.sum = &kernel_sum.sum + &kernel.excess;
+            }
+
+            if max_kernel_timelock > height {
+                return Err(ValidationError::MaturityError);
+            }
+
+            if coinbase_index.is_none() {
+                warn!(
+                    target: LOG_TARGET,
+                    "Block #{} failed to validate: no coinbase kernel", height
+                );
+                return Err(ValidationError::TransactionError(TransactionError::NoCoinbase));
+            }
+
+            let coinbase_index = coinbase_index.unwrap();
+
+            Ok(KernelValidationData {
+                kernels,
+                kernel_sum,
+                coinbase_index,
+            })
+        })
+        .into()
+    }
+
+    fn start_input_validation(
+        &self,
+        header: &BlockHeader,
+        output_hashes: Vec<HashOutput>,
+        inputs: Vec<TransactionInput>,
+    ) -> AbortOnDropJoinHandle<Result<InputValidationData, ValidationError>> {
+        let block_height = header.height;
+        let commitment_factory = self.factories.commitment.clone();
+        let db = self.db.inner().clone();
+        task::spawn_blocking(move || {
+            let mut aggregate_input_key = PublicKey::default();
+            let mut commitment_sum = Commitment::default();
+            let mut not_found_inputs = Vec::new();
+            let db = db.db_read_access()?;
+            for (i, input) in inputs.iter().enumerate() {
+                // Check for duplicates and/or incorrect sorting
+                if i > 0 && input <= &inputs[i - 1] {
+                    return Err(ValidationError::UnsortedOrDuplicateInput);
+                }
+
+                if !input.is_mature_at(block_height) {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Input found that has not yet matured to spending height: {}", block_height
+                    );
+                    return Err(TransactionError::InputMaturity.into());
+                }
+
+                match helpers::check_input_is_utxo(&*db, input) {
```

Comment:
I think the header hash  perhaps previous_hash should be passed in here in the case where a block is added or removed from the chain while the validator is running
```suggestion
                match helpers::check_input_is_utxo(&*db, input, header.previous_hash) {
```

Reply:
Although I see the `db_read_access` above, so this is less of a worry, but a block may still be added before this task starts

Reply:
Yeah it's very unlikely, but possible if external code adds propagated blocks when the state machine is in block sync state. You fundamentally cannot share a single lock across threads, so we'll need to introduce that invariant here somehow either by ignoring propagated blocks while the base node is in sync mode as a rule / set a flag in the db so disable calls to `write` (i.e. read only mode). Open to suggestions.

Reply:
As you suggested, atomically checking the header hash vs the current header hash should also work. I'll add that as a final check before committing the new block.

Reply:
Why not just check the height or hash of the tip after acquiring the read lock?

Reply:
We could do that, but that will only save us starting an unnecessary validation that will fail. The main concern is when you release the read lock, post validation and acquire the write lock to commit the block - where it is possible for a data race: external code waiting on a write lock to write a different valid block to the tip of the chain, obtains the lock writes block A' to the tip, post validation acquires the write lock and writes the new tip. Come to think of it, I think we already prevent this on the database layer when inserting headers